### PR TITLE
Updated Schema, added route + controller function for finding ID + tests.

### PR DESCRIPTION
**Pull request by Heyjp**
Need some guidance here. 

Tried out using Jest but couldn't get environment variables to work so hard coded the local IP. Is there any way to do this? Also with the Schema, why are we wrapping _id in an object as oppose to letting Mongoose handle it. Or even by just writing . 

```
{
  _id: ObjectId
}
```

### CODE REVIEW

**Comment by zsoltime** [comment]
I can review your PR when I'm back at home today.

The schema I posted on #40 has an `_id` field which is an ObjectID generated by MongoDB. Although never used it, I think Mongoose uses the same ObjectID when it inserts a document. 

The seed files I uploaded are seed files only and we can use with `mongoimport`. The `_id` `$oid` fields are becoming `"_id" : ObjectId("xxxxxxx")` by importing these JSON files. It might be confusing if you've never used `mongoexport` to export data.

File: server/controller/index.js [hunk 1]
```diff
@@ -25,3 +25,16 @@ exports.getQuestions = async (ctx) => {
       return ctx.body;
     });
 };
+
+exports.getId = async (ctx) => {
+  console.log(ctx.params.id, "this is params")
```
Comment (zsoltime) [inline server/controller/index.js]: You should use single quotes.

File: server/controller/index.js [hunk 2]
```diff
@@ -25,3 +25,16 @@ exports.getQuestions = async (ctx) => {
...
+  await Question.findById(ctx.params.id)
+    .then((result) => {
+      ctx.body = result;
+      return ctx.body;
+    })
+    .catch((err) => {
+      ctx.body = "id not recognized";
```
Comment (zsoltime) [inline server/controller/index.js]: You should use single quotes.

File: server/tests/testSchema.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
```
Comment (zsoltime) [inline server/tests/testSchema.js]: You should use single quotes.

File: server/models/question.js [hunk 4]
```diff
@@ -1,11 +1,35 @@
 const mongoose = require('mongoose');
 
 const questionSchema = mongoose.Schema({
+  _id: {
```
Comment (zsoltime) [inline server/models/question.js]: Not sure if we need the `_id` field here as I mentioned in my [previous post](https://github.com/chingu-voyage3/bears-17/pull/43#issuecomment-351133803).

File: server/models/question.js [hunk 5]
```diff
@@ -1,11 +1,35 @@
...
+    $oid: ObjectId,
+  },
+  submitted_at: Date.now(),
```
Comment (zsoltime) [inline server/models/question.js]: In your previous PR you set `timestamps: true`. Not sure which one is the preferred way, but we can use `timestamps: { createdAt: 'submitted_at', updatedAt: 'updated_at' }` to change the default names.

File: server/models/question.js [hunk 6]
```diff
@@ -1,11 +1,35 @@
...
+  body: String,
+  votes: Number,
+  author: {
+    _id: {
+      $oid: ObjectId()
+    },
+    name: String,
+    avatar: String,
+  }, 
+  answers: [{
```
Comment (zsoltime) [inline server/models/question.js]: We can remove answers now as it has its own collection (PR #42)

**Review by zsoltime — CHANGES_REQUESTED** [review]
Good job, however, I haven't check the tests yet. You should rename your `test.js` file to something more meaningful, like `questions.test.js` or something like that?

Could you set up your editor to trim trailing spaces and add a final newline? Also, not sure if you use the ESLint plugin as it should "scream" when you use double quotes, miss commas, etc.

**Comment by zsoltime** [comment]
Just noticed that I forgot to answer your very first question. To get the content of the `.env` file, you need to add `require('dotenv').config();` on the first line.

However, we should to create a new database first for testing purposes, so tests don't wipe the db we use for development at the moment.

**Comment by Heyjp** [comment]
We can create a new DB that is an option, or we can just simply change the collection that we use to store and remove all the data. 

```
const questionSchema = mongoose.Schema(
  {
    title: String,
    question: String,
    user: { default: 'user', type: String },
  },
  {
    timestamps: true,
    collection: 'test',
  },
);
```

Currently the tests run off a local instance of mongodb, on the 'test' collection as well so I don't think it should affect the production DB at least. Not sure if this is the best option happy to be corrected though.

File: server/models/question.js [hunk 7]
```diff
@@ -1,12 +1,23 @@
...
-});
+const questionSchema = mongoose.Schema(
+  {
+    submitted_at: Date.now(),
+    title: String,
+    body: String,
+    votes: Number,
+    author: {
+      _id: {
+        $oid: ObjectId(),
```
Comment (zsoltime) [inline server/models/question.js]: `ObjectId` is not defined and maybe we should add a new `author` schema like in the [docs](http://mongoosejs.com/docs/subdocs.html)?

File: server/models/question.js [hunk 8]
```diff
@@ -1,12 +1,23 @@
...
-const questionSchema = mongoose.Schema({
-  title: String,
-  question: String,
-  user: { default: 'user', type: String },
-}, {
-  timestamps: true,
-});
+const questionSchema = mongoose.Schema(
+  {
+    submitted_at: Date.now(),
```
Comment (zsoltime) [inline server/models/question.js]: Do we still need this one? You've added timestamps already.

File: server/models/question.js (hunk 7 above)
Comment (Heyjp) [reply server/models/question.js]: I agree with this. Would make the DB easier to manage.

Comment (Heyjp) [reply server/models/question.js]: I think we can add in a new schema for the `author` but we still need to keep a reference in the `question` schema. Would a string be better or should we just use an `ObjectId` that we would get from the new schema?

Comment (zsoltime) [reply server/models/question.js]: The author schema should be no more than a simple `{ name: String, avatar: String }` we don't need to add the `_id` field here, at least I don't think so. But I don't know mongoose :( Could you test it this way?

**Comment by Heyjp** [comment]
So made a slight alteration to the schemas so that we can collect information across both of them if we so choose. This could be useful for say inspecting a profile and wanting to see maybe a list of questions / answers that the user has made. 

Followed the same format as the json files uploaded. Hope my formatting is on point this time.

**Comment by zsoltime** [comment]
So is your `authorSchema` the same as `userSchema` would be? I really need to dive into reading more about mongoose population and subdocs, but if I'm correct subdocs are embedded documents (meaning there are no extra queries to get the user's name and avatar) while populated queries are running extra queries behind the scenes.

So if we list 50 questions on a page and we want to display the question title, the date, the author name and avatar, we need 1 query to get the question and 50 more queries to get the author name and avatar. 

We can avoid it by embedding the author name and avatar inside the question and use one single query instead of the 51 above.

**Comment by Heyjp** [comment]
Changed the question to have an embedded author like you said. But also removed the questions on the `userSchema` / `authorSchema`. I think that could still benefit from populate on that schema, otherwise you are just making another query to the DB to retrieve all the questions a user might have. Could expand to have Answers as well.

**Comment by zsoltime** [comment]
I think you're right, but let's just do that in one of the next sprints and leave it as is. It works as expected and all tests pass. There are a few styling issues I'd like you to fix before I can accept and merge this. I'll comment these in a few minutes.

Anyway, really great job, Jamie, thank you. 👍 😃

File: server/models/question.js [hunk 9]
```diff
@@ -1,12 +1,19 @@
...
+    title: { type: String, required: true },
+    body: String,
+    votes: Number,
+    author: {
+      name: { type: String, required: true },
+      avatar: String,
+    },
+  },
+  {
+    timestamps: { createdAt: 'submitted_at', updatedAt: 'updated_at' }
```
Comment (zsoltime) [inline server/models/question.js]: There is a missing comma

File: server/models/user.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+const mongoose = require('mongoose');
+
+const userSchema = mongoose.Schema({
+  name: { type: String, required: true },
+  email: String,
+  avatar: String,
+  country: String,
+  member_since: {type: Date, default: Date.now },
```
Comment (zsoltime) [inline server/models/user.js]: There is a missing space before `type`

File: server/tests/questions.test.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    await Question.find({})
+      .then((result) => {
+        expect(result).toHaveLength(1);
+        expect(result[0].title).toEqual(questionData.title);
+      });
+
+    done();
+  });
+  
+  test('Should remove a single entry and return an empty array ', async (done) => {
```
Comment (zsoltime) [inline server/tests/questions.test.js]: There is an extra space after `array`

File: server/tests/questions.test.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      .then(result => result[0].remove()).then(() => {
+        Question.find({ _id: questionData._id })
+          .then((res) => {
+            expect(res).toHaveLength(0);
+          });
+      });
+    done();
+  });
+
+  test('should find and return an entry based on id', async (done) => {
```
Comment (zsoltime) [inline server/tests/questions.test.js]: `should` should be capitalised (or all the other ones lowercase 😃) to stay consistent.

File: server/tests/testSchema.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+);
+
+const questionModel = mongoose.model('Question', questionSchema);
+const userModel = mongoose.model('User', userSchema);
+
+module.exports = {
+  User: userModel,
+  Question: questionModel,
+};
+
```
Comment (zsoltime) [inline server/tests/testSchema.js]: There are two new lines here, should be one only.

**Review by zsoltime — CHANGES_REQUESTED** [review]

**Review by zsoltime — APPROVED** [review]

**Pull request merged**